datanode/api/trading_data_v2.go

```diff
@@ -2632,16 +2644,198 @@ func (t *TradingDataServiceV2) estimateMargin(
+		SearchLevel:            maintenanceMargin.Mul(num.DecimalFromFloat(scalingFactors.SearchLevel)).Round(0).String(),
+		InitialMargin:          maintenanceMargin.Mul(num.DecimalFromFloat(scalingFactors.InitialMargin)).Round(0).String(),
+		CollateralReleaseLevel: maintenanceMargin.Mul(num.DecimalFromFloat(scalingFactors.CollateralRelease)).Round(0).String(),
+	}
+}
+
+func (t *TradingDataServiceV2) EstimatePosition(ctx context.Context, req *v2.EstimatePositionRequest) (*v2.EstimatePositionResponse, error) {
+	defer metrics.StartAPIRequestAndTimeGRPC("EstimatePosition")()
+
+	rf, err := t.riskFactorService.GetMarketRiskFactors(ctx, req.MarketId)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting risk factors: %s", req.MarketId)
+	}
+
+	mkt, err := t.marketService.GetByID(ctx, req.MarketId)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting market: %s", req.MarketId)
+	}
+
+	mktData, err := t.marketDataService.GetMarketDataByID(ctx, req.MarketId)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting market data: %s", req.MarketId)
+	}
+
+	mktProto := mkt.ToProto()
+
+	asset, err := mktProto.GetAsset()
+	if err != nil {
+		return nil, errors.Wrap(err, "getting asset from market")
+	}
+
+	marketObservable, err := num.DecimalFromString(mktData.MarkPrice.String())
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing mark price: %s", mktData.MarkPrice.String())
+	}
+
+	auction := mktData.AuctionEnd > 0
+	if auction && mktData.MarketTradingMode == types.MarketTradingModeOpeningAuction.String() {
+		marketObservable = mktData.IndicativePrice
+	}
+
+	priceFactor, err := t.getMarketPriceFactor(ctx, mkt)
+	if err != nil {
+		return nil, err
+	}
+
+	dPriceFactor := priceFactor.ToDecimal()
+
+	uMarketObservable, overflowed := num.UintFromDecimal(marketObservable)
+	if overflowed {
+		return nil, fmt.Errorf("mark price overflowed: %s", mktData.MarkPrice)
+	}
+
+	marketObservableInAssetPrecision, err := t.scaleFromMarketToAssetPrice(ctx, mkt, uMarketObservable)
+	if err != nil {
+		return nil, errors.Wrap(ErrScalingPriceFromMarketToAsset, err.Error())
+	}
+
+	dMarketObservableInAssetPrecision := num.DecimalFromUint(marketObservableInAssetPrecision)
+
+	positionFactor := num.DecimalFromFloat(10).
+		Pow(num.DecimalFromInt64(int64(mkt.PositionDecimalPlaces)))
+
+	linearSlippageFactor, err := num.DecimalFromString(mktProto.LinearSlippageFactor)
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing linear slippage factor: %s", mktProto.LinearSlippageFactor)
+	}
+	quadraticSlippageFactor, err := num.DecimalFromString(mktProto.QuadraticSlippageFactor)
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing quadratic slippage factor: %s", mktProto.QuadraticSlippageFactor)
+	}
+
+	buyOrders := make([]*risk.OrderInfo, 0, len(req.Orders))
+	sellOrders := make([]*risk.OrderInfo, 0, len(req.Orders))
+
+	for _, o := range req.Orders {
```

Comment:
Should we check that `len(req.Orders) != 0` right at the top of the function call? I'm imagining this would be an `InvalidArgument` type of error.

Reply:
`len(req.Orders)` == 0 is fine, party may not have any open. Added a check for nil order.